product_variant_default_code/tests/test_variant_default_code.py

```diff
@@ -208,12 +221,93 @@ def test_09_code_change_propagation(self):
+                + "-"
+                + product.product_template_attribute_value_ids.filtered(
+                    lambda x: x.product_attribute_value_id.attribute_id == self.attr2
+                ).name[0:2]
+            )
+            product.flush()
+            self.assertEqual(product.default_code, expected_code)
+
+    def test_13_new_attribute(self):
+        self.assertEqual(self.template1.reference_mask, "[TSize]-[TColor]")
+        self.assertEqual(len(self.template1.mapped("product_variant_ids")), 4)
+
+        self.attr3 = self.env["product.attribute"].create({"name": "TCollection"})
+        self.attr3_1 = self.env["product.attribute.value"].create(
+            {"name": "New", "attribute_id": self.attr3.id}
+        )
+        self.attr3_2 = self.env["product.attribute.value"].create(
+            {"name": "Old", "attribute_id": self.attr3.id}
+        )
+
+        self.template1.write(
+            {
+                "attribute_line_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "attribute_id": self.attr3.id,
+                            "value_ids": [(6, 0, [self.attr3_1.id, self.attr3_2.id])],
+                        },
+                    ),
+                ]
+            }
+        )
+
+        self.assertEqual(
+            self.template1.reference_mask, "[TSize]-[TColor]-[TCollection]"
+        )
+        self.assertEqual(len(self.template1.mapped("product_variant_ids")), 8)
+
+        for product in self.template1.mapped("product_variant_ids"):
+            expected_code = (
+                product.product_template_attribute_value_ids.filtered(
+                    lambda x: x.product_attribute_value_id.attribute_id == self.attr1
+                ).name[0:2]
+                + "-"
+                + product.product_template_attribute_value_ids.filtered(
+                    lambda x: x.product_attribute_value_id.attribute_id == self.attr2
+                ).name[0:2]
+                + "-"
+                + product.product_template_attribute_value_ids.filtered(
+                    lambda x: x.product_attribute_value_id.attribute_id == self.attr3
+                ).name[0:2]
+            )
+            self.assertEqual(product.default_code, expected_code)
+
+    def test_14_rename_attribute(self):
+        self.assertEqual(self.template1.reference_mask, "[TSize]-[TColor]")
+        self.attr1.name = "TNewSize"
+        self.attr1.onchange_default_code_attribute()
+        self.assertEqual(self.template1.reference_mask, "[TNewSize]-[TColor]")
+
+    def test_15_sequence_change(self):
+        self.assertEqual(self.template1.reference_mask, "[TSize]-[TColor]")
+        self.attr1.sequence = 1
+        self.attr2.sequence = 0
+        self.template1.reference_mask = "[TColor]-[TSize]"
+        self.template1.write({"name": "New"})
+        self.assertEqual(self.template1.reference_mask, "[TColor]-[TSize]")
```

Comment:
TODO
- add a test where you change a code of a product.attribute
- add a test where you change a code of a product.attribute.value

Reply:
Test 9 and 10 already test them, I just add the code change of an active product.attribute.value in test 9

Reply:
ok